multicluster/service-mirror/cluster_watcher.go

```diff
@@ -655,6 +683,38 @@ func (rcsw *RemoteClusterServiceWatcher) Start(ctx context.Context) error {
 			},
 		},
 	)
+	// Add an event handler for Endpoints object to see whether we need to
+	// update mirrored headless services. We only care if an Endpoints is
+	// updated and if it belongs
+	// to a headless service.
+	rcsw.remoteAPIClient.Endpoint().Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(old, new interface{}) {
+				if old.(v1.Object).GetNamespace() == "kube-system" {
+					return
+				}
+
+				ep, ok := new.(*corev1.Endpoints)
+				if !ok {
+					rcsw.log.Errorf("error processing Endpoints resource, got %#v, expected *corev1.Endpoints", new)
+					return
+				}
+
+				// If we have at least 1 Ready Address process the object.
+				if len(ep.Subsets[0].Addresses) == 0 {
+					return
+				}
+
+				if _, found := ep.Labels["service.kubernetes.io/headless"]; !found {
```

Comment:
For now, I left everything as is (except for using a shared informer in `createOrUpdateHeadlessEndpoints()`, however I think you raise a valid point of checking for the export label.

> in case the user decides to unexport something and we'd have to clean things up in the source cluster. If we don't do it here, we'd have to introduce that logic again in the Service listener.

I added in a function to clean up clusterIP services and it is done based on the local Endpoints object. However, we're not dealing with any remote Endpoints object. If I understand correctly, we should be able to only let exported Endpoints through in the handler -- in the clean up function we query local Endpoints based on the serviceNamespace/serviceName identifier. Think even if a user would unexport their service (and hence the Endpoints wouldn't be picked up) we could still clean up properly. Wdyt?

Reply:
Yeah, cleaning up endpoints on the service handlers sounds good to me :+1: